include/envoy/registry/registry.h

```diff
@@ -388,9 +388,16 @@ template <class Base> class FactoryRegistry : public Logger::Loggable<Logger::Id
     auto config_type = factory.configType();
     Base* prev = getFactoryByType(config_type);
     if (prev != nullptr) {
-      factoriesByType().emplace(config_type, &factory);
+      ASSERT(displaced == nullptr || displaced == prev,
+             fmt::format("displaced two different factories: one by name ({}) and one by type ({})",
+                         factory.name(), config_type));
+      displaced = prev;
+      factoriesByType().erase(config_type);
     }
 
+    factoriesByType().emplace(config_type, &factory);
```

Comment:
I think I put at guard here before to deal with the double-registered factories. The invariant is that two factories that have the same config type should not be present in the config type registry. I don't think we need to maintain that invariant for test injected factories.

Reply:
I rewrote replaceFactoryForTest to completely remove the displaced factory by name and type (and to log displacements). Had to fix a couple tests that used the same type for multiple factories in one test, but otherwise I think it should be pretty bullet proof now. We'll see how the tests go.